### PR TITLE
DIT-961 Reporting Endpoint

### PR DESCRIPTION
**Pull request by ed0906**

### CODE REVIEW

File: test/util/Cases.scala [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+package util
+
+import java.time.Instant
+import java.util.UUID
+
+import uk.gov.hmrc.bindingtariffclassification.model.CaseStatus.CaseStatus
+import uk.gov.hmrc.bindingtariffclassification.model._
+
+object Cases {
```
Comment (ed0906) [inline test/util/Cases.scala]: Copied from the internal UI. A pretty useful case builder.

File: app/uk/gov/hmrc/bindingtariffclassification/model/CaseFilter.scala [hunk 2]
```diff
@@ -55,25 +53,9 @@ object CaseFilter {
...
-
-      def params(name: String): Option[Set[String]] = {
-        requestParams.get(name).map(_.flatMap(_.split(",")).toSet).filter(_.exists(_.nonEmpty))
-      }
-
-      def param(name: String): Option[String] = {
-        params(name).map(_.head)
-      }
+      import uk.gov.hmrc.bindingtariffclassification.model.utils.BinderUtil._
+      implicit val rp: Map[String, Seq[String]] = requestParams
```
Comment (retroredge) [inline app/uk/gov/hmrc/bindingtariffclassification/model/CaseFilter.scala]: can this line live inside the BinderUtil class?

File: app/uk/gov/hmrc/bindingtariffclassification/model/reporting/CaseReport.scala [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  private val reportFieldKey = "report_field"
+
+  implicit def bindable(implicit stringBinder: QueryStringBindable[String],
+                        filterBinder: QueryStringBindable[CaseReportFilter]
+                     ): QueryStringBindable[CaseReport] = new QueryStringBindable[CaseReport] {
+    override def bind(key: String, requestParams: Map[String, Seq[String]]): Option[Either[String, CaseReport]] = {
+      import uk.gov.hmrc.bindingtariffclassification.model.utils.BinderUtil._
+      implicit val rp: Map[String, Seq[String]] = requestParams
+
+      val filter: CaseReportFilter = filterBinder.bind("", requestParams).get.right.get
```
Comment (SteveSugden) [inline app/uk/gov/hmrc/bindingtariffclassification/model/reporting/CaseReport.scala]: is get safe here?   No idea what filterBinder.bind returns

File: app/uk/gov/hmrc/bindingtariffclassification/model/reporting/CaseReportFilter.scala [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  val decisionStartKey = "decision_start"
+
+  implicit def binder(implicit rangeBinder: QueryStringBindable[InstantRange]): QueryStringBindable[CaseReportFilter]
+  = new QueryStringBindable[CaseReportFilter] {
+
+    override def bind(key: String, requestParams: Map[String, Seq[String]]): Option[Either[String, CaseReportFilter]] = {
+      implicit val rp: Map[String, Seq[String]] = requestParams
+
+      val decisionStart: Option[InstantRange] = rangeBinder.bind(decisionStartKey, requestParams).filter(_.isRight).map(_.right.get)
+
```
Comment (SteveSugden) [inline app/uk/gov/hmrc/bindingtariffclassification/model/reporting/CaseReportFilter.scala]: The use of 'get' looks suspicious here as well

File: app/uk/gov/hmrc/bindingtariffclassification/controllers/ReportingController.scala [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.bindingtariffclassification.controllers
+
+import javax.inject.{Inject, Singleton}
+import uk.gov.hmrc.bindingtariffclassification.service.ReportService
+
+@Singleton
+class ReportingController @Inject()(reportService: ReportService) extends CommonController {
```
Comment (pllaurado) [inline app/uk/gov/hmrc/bindingtariffclassification/controllers/ReportingController.scala]: ???

File: app/uk/gov/hmrc/bindingtariffclassification/service/ReportService.scala [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+package uk.gov.hmrc.bindingtariffclassification.service
+
+import javax.inject.Inject
+import uk.gov.hmrc.bindingtariffclassification.model.reporting.{CaseReport, ReportResult}
+import uk.gov.hmrc.bindingtariffclassification.repository.CaseRepository
+
+import scala.concurrent.Future
+
+class ReportService @Inject()(caseRepository: CaseRepository) {
```
Comment (retroredge) [inline app/uk/gov/hmrc/bindingtariffclassification/service/ReportService.scala]: A general comment about our services in this app is that they seem very 'anemic' and the methods typically do nothing expect delegate to the repos. Maybe an indication that we're doing too much in the controller or repo layer

File: test/unit/uk/gov/hmrc/bindingtariffclassification/model/reporting/CaseReportTest.scala [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+  private val params: Map[String, Seq[String]] = Map(
+    "min_decision_start" -> Seq("1970-01-01T00:00:00Z"),
+    "max_decision_start" -> Seq("1970-01-01T00:00:01Z"),
+    "report_field" -> Seq("days-elapsed"),
+    "report_group" -> Seq("queue-id")
+  )
+
+  /**
+    * When we add fields to Report these tests shouldn't need changing, only the fields above.
```
Comment (retroredge) [inline test/unit/uk/gov/hmrc/bindingtariffclassification/model/reporting/CaseReportTest.scala]: nice

File: app/uk/gov/hmrc/bindingtariffclassification/controllers/ReportingController.scala (hunk 5 above)
Comment (ed0906) [reply app/uk/gov/hmrc/bindingtariffclassification/controllers/ReportingController.scala]: ??? to you too @pllaurado

File: app/uk/gov/hmrc/bindingtariffclassification/model/reporting/CaseReport.scala (hunk 3 above)
Comment (ed0906) [reply app/uk/gov/hmrc/bindingtariffclassification/model/reporting/CaseReport.scala]: If you have a look in the Filter binder, it always returns `Some(Right(...))` so it's safe

File: app/uk/gov/hmrc/bindingtariffclassification/model/reporting/CaseReportFilter.scala (hunk 4 above)
Comment (ed0906) [reply app/uk/gov/hmrc/bindingtariffclassification/model/reporting/CaseReportFilter.scala]: Similar to above. The InstantRange binder always returns `Some(Right(...))`

Comment (ed0906) [reply app/uk/gov/hmrc/bindingtariffclassification/model/reporting/CaseReportFilter.scala]: Tbh would probably be better to pass up any errors as `Left`s

File: app/uk/gov/hmrc/bindingtariffclassification/service/ReportService.scala (hunk 6 above)
Comment (ed0906) [reply app/uk/gov/hmrc/bindingtariffclassification/service/ReportService.scala]: Yep good point, I don't think there's a whole lot of business logic in this Microservice (intentionally) which usually belongs in the Service Layer. The only logic we have really is the case reference generation. Not sure what controller/repo code we could pull into the services though really, everything seems in the right place already to me

File: app/uk/gov/hmrc/bindingtariffclassification/model/CaseFilter.scala (hunk 2 above)
Comment (ed0906) [reply app/uk/gov/hmrc/bindingtariffclassification/model/CaseFilter.scala]: I tried, the request params would have to be marked as implicit which causes other issues

**Pull request merged**